### PR TITLE
Hypre AMG on GPU

### PR DESCRIPTION
**Pull request by rrsettgast**

### CODE REVIEW

**Comment by rrsettgast** [comment]
@bmhan12 Can you help me out with this PR? I am getting an error on the `gcc-cuda` build:

```_hypre_utilities.hpp:27:10: fatal error: cublas_v2.h: No such file or directory
3184 #include <cublas_v2.h>
3185
```

This is because the file doesn't exist in the cuda path. It appears to exist in the 10.2 path, but I am don't understand why that is there. I think there was some tricks you had to play to get this image working. I think it would be better to have you fix it. 

Thanks
Randy

**Comment by bmhan12** [comment]
@rrsettgast Interesting... I'll look into it.

**Comment by bmhan12** [comment]
@rrsettgast  The Docker image builds locally for me but not on Travis. Looking into a fix.

**Comment by rrsettgast** [comment]
@corbett5 Can you review

File: docker/gcc-cuda/Dockerfile [hunk 1]
```diff
@@ -118,6 +121,10 @@ ARG CUDA_TOOLKIT_ROOT_DIR=/usr/local/cuda
 ARG CUDA_ARCH=sm_70
 ARG CMAKE_CUDA_COMPILER=${CUDA_TOOLKIT_ROOT_DIR}/bin/nvcc
 
+ENV HYPRE_CUDA_SM=70
+ENV CUDA_HOME=${CUDA_TOOLKIT_ROOT_DIR}
+#RUN ln -s ln -s /usr/include/cublas_v2.h /usr/local/cuda/include/cublas_v2.h
+
```
Comment (rrsettgast) [inline docker/gcc-cuda/Dockerfile:127]:
```suggestion
```

File: docker/gcc-cuda/Dockerfile [hunk 2]
```diff
@@ -57,6 +59,7 @@ RUN yum install -y \
 
 ENV LIBRARY_PATH /usr/local/cuda/lib64/stubs
 
+RUN yum install -y yum-plugin-versionlock && yum versionlock libcublas10
```
Comment (rrsettgast) [inline docker/gcc-cuda/Dockerfile:62]:
```suggestion
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -83,6 +90,33 @@ message(STATUS "Build command = ${TPL_BUILD_COMMAND}")
 include(GNUInstallDirs)
 include( cmake/thirdparty/FindMathLibraries.cmake )
 
+
+set( GEOSX_LOCALINDEX_TYPE "std::ptrdiff_t" CACHE STRING "" )
+set( GEOSX_GLOBALINDEX_TYPE "long long int" CACHE STRING "" )
+
+if( GEOSX_LOCALINDEX_TYPE STREQUAL "int" )
```
Comment (corbett5) [inline CMakeLists.txt:97]: What is the point of this conversion?

**Review by corbett5 — APPROVED** [review]

File: CMakeLists.txt (hunk 3 above)
Comment (rrsettgast) [reply CMakeLists.txt:97]: We have to support `int` index types for hypre on the gpu....so fun.

Comment (corbett5) [reply CMakeLists.txt:97]: Yeah I get that, I'm just wondering what the point of the `GEOSX_LOCALINDEX_TYPE_FLAG` is, the conversion seems unnecessary and complicated. Like how am I supposed to know what `3` means. Why not just keep the string?

**Pull request merged**